kore/src/Kore/Parser/Lexer.x

```diff
@@ -0,0 +1,328 @@
+  return = App.pure
+
+alexGetByte :: AlexInput -> Maybe (Word8, AlexInput)
+alexGetByte (AlexInput {alexPosn=p,alexStr=cs,alexBytePos=n}) =
+    case ByteString.uncons cs of
+        Nothing -> Nothing
+        Just (c, rest) -> 
+          let b = ByteString.w2c c in
+            Just (c, AlexInput {
+                alexPosn = alexMove p b,
+                alexChar = b,
+                alexStr =  rest,
+                alexBytePos = n+1})
+
+alexMove :: AlexPosn -> Char -> AlexPosn
+alexMove (AlexPn fp a l c) '\t' =
+    AlexPn fp (a+1) l (c+alex_tab_size-((c-1) `mod` alex_tab_size))
+alexMove (AlexPn fp a l _) '\n' = AlexPn fp (a+1) (l+1) 1
+alexMove (AlexPn fp a l c) _    = AlexPn fp (a+1) l (c+1)
+
+alexError :: FilePath -> Int -> Int -> String -> Alex a
+alexError fp line column msg = do
+    Alex
+    $ const
+    $ Left
+    $ fp ++ ":" ++ show line ++ ":" ++ show column ++ ": " ++ msg ++ "\n"
+
+-- End Alex wrapper code.
+
+{- | Convert the textual representation of a string token into its semantic
+value as a Text. Returns Left with an error message if it fails, otherwise
+Right.
+-}
+unescape :: Text -> Either String Text
+unescape t =
+  let dropQuotes str = Text.take (Text.length str - 2) . Text.drop 1 $ str
+      escape rest c = either Left (\s -> Right (c : s)) $ go rest
+      validate rest c
+          | c >= 0x10ffff = Left ("code point " ++ show c ++ " outside range of Unicode")
+          | generalCategory (chr c) == Surrogate 
+              = Left ("surrogate character " ++ show (chr c) ++ " in string literal")
+          | otherwise = escape rest $ chr c
+      go [] = Right []
+      go ('\\' : 'n' : rest) = escape rest '\n'
+      go ('\\' : 'r' : rest) = escape rest '\r'
+      go ('\\' : 't' : rest) = escape rest '\t'
+      go ('\\' : 'f' : rest) = escape rest '\f'
+      go ('\\' : '\\' : rest) = escape rest '\\'
+      go ('\\' : '"' : rest) = escape rest '"'
+      go ('\\' : 'x' : first : second : rest) =
+          escape rest $ chr (digitToInt first * 16 +
+                             digitToInt second)
+      go ('\\' : 'u' : first : second : third : fourth : rest) = 
+          validate rest $ (((digitToInt first * 16 + 
+                             digitToInt second) * 16 + 
+                             digitToInt third) * 16 + 
+                             digitToInt fourth)
+      go ('\\' : 'U' : first : second : third : fourth
+                     : fifth : sixth : seventh : eighth : rest) =
+          validate rest $ (((((((digitToInt first * 16 + 
+                                 digitToInt second) * 16 + 
+                                 digitToInt third) * 16 + 
+                                 digitToInt fourth) * 16 + 
+                                 digitToInt fifth) * 16 +
+                                 digitToInt sixth) * 16 +
+                                 digitToInt seventh) * 16 + 
+                                 digitToInt eighth)
```

Comment:
In an effort to merge cases and simplify logic here, how about something like this:
```suggestion
      go ('\\' : pre : rest0)
        | Just size <- lookup pre $ zip "xuU" [2,4,8]
        , (body, rest) <- splitAt size rest0
        = runReadS Numeric.readHex body
            >>= validate rest
      where
          runReadS reader str = case reader str of
              [(num, "")] -> pure num
              _other -> Left "invalid hex literal"
              -- probably can be improved by moving function into Alex
```
This will require a `import qualified Data.Numeric as Numeric`. Additionally note that it calls `validate` in the `\xHH` case, which seems like that should be OK. If this causes a big slowdown then that case can be factored back out.